### PR TITLE
Pebryan/2020 2 16 d365

### PR DESCRIPTION
**Pull request by petebryan**
New Hunting Queries and Detections for Dynamics 365 connector.

Detections:
- User Bulk Retrieval Outside Normal Activity
- New Dynamics User Agents
- New Office User Agents in Dynamics
- Mass Export from Dynamics to Excel
- New Admin Activity
- Dynamics Encryption Setting Change

Hunting Queries:
- Dynamics Activity after Failed Logons
- Dynamics Activity after AAD Alert

### CODE REVIEW

File: Hunting Queries/Dynamics365Activity/DynamicsActivityAfterFailedLogons.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  SigninLogs
+  | where ResultType in ("50125", "50140", "70043", "70044")
+  | summarize count() by IPAddress
+  | where count_ >= threshold
+  | join (Dynamics365Activity
+  | extend IPAddress = tostring(split(ClientIP, ":")[0]))
+  on IPAddress
+  | project-rename FailedLogonCount = count_
+  | extend timestamp = TimeGenerated, AccountCustomEntity = UserId, IPCustomEntity = IPAddress
+
```
Comment (shainw) [inline Hunting Queries/Dynamics365Activity/DynamicsActivityAfterFailedLogons.yaml:24]: Missing Entity mappings, this is now required...even for Hunting. ;)

File: Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+relevantTechniques:
+  - T1530
+query: |
+  let baseline_time = 7d;
+  let detection_time = 1d;
+  Dynamics365Activity
+  | where TimeGenerated between(ago(baseline_time)..ago(detection_time-1d))
+  | where OriginalObjectId contains 'ExportToExcel'
+  | extend numQueryCount = todouble(QueryResults)
+  | extend QueryCount = iif(QueryResults contains ",", todouble(countof(tostring(QueryResults), ',') + 1), numQueryCount)
```
Comment (shainw) [inline Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml:25]: I don't have a test environment with this, but would it have better perf if we use array_length here?  Looks like QueryResult is already a dynamic?

File: Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  | extend QueryCount = iif(isnotempty(QueryCount), QueryCount, double(1))
+  | summarize sum(QueryCount) by UserId
+  | extend HistoricalBaseline = sum_QueryCount
+  | join (Dynamics365Activity
+  | where TimeGenerated > ago(detection_time)
+  | where OriginalObjectId contains 'ExportToExcel'
+  | extend numQueryCount = todouble(QueryResults)
+  | extend QueryCount = iif(QueryResults contains ",", todouble(countof(tostring(QueryResults), ',') + 1), numQueryCount)
```
Comment (shainw) [inline Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml:33]: Same as above on array_length

File: Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  | extend QueryCount = iif(isnotempty(QueryCount), QueryCount, double(1))
+  | summarize sum(QueryCount) by UserId
+  | extend CurrentExportRate = sum_QueryCount) on UserId
+  | where CurrentExportRate > HistoricalBaseline
+  | project UserId, HistoricalBaseline, CurrentExportRate
+  | join kind=inner(Dynamics365Activity
+  | where TimeGenerated > ago(detection_time)
+  | where OriginalObjectId contains 'ExportToExcel'
+  | extend numQueryCount = todouble(QueryResults)
+  | extend QueryCount = iif(QueryResults contains ",", todouble(countof(tostring(QueryResults), ',') + 1), numQueryCount)
```
Comment (shainw) [inline Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml:43]: Same as above on array_length

File: Detections/Dynamics365Activity/NewDynamicsUserAgent.yaml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  let known_useragents = dynamic([]);
+  Dynamics365Activity
+  | where TimeGenerated between(ago(lookback)..ago(timeframe))
+  | where isnotempty(UserAgent)
+  | summarize by UserAgent, UserId
+  | join kind = rightanti (Dynamics365Activity
+  | where TimeGenerated > ago(timeframe)
+  | where isnotempty(UserAgent)
+  | where UserAgent !in (known_useragents)
+  | where UserAgent !startswith "azure-logic-apps"
```
Comment (shainw) [inline Detections/Dynamics365Activity/NewDynamicsUserAgent.yaml:31]: Startswith is more expensive, is it possible to use "has" here?

File: Detections/Dynamics365Activity/NewDynamicsUserAgent.yaml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  let user_accounts = "(([a-zA-Z]{1,})\\.([a-zA-Z]{1,}))@.*";
+  let known_useragents = dynamic([]);
+  Dynamics365Activity
+  | where TimeGenerated between(ago(lookback)..ago(timeframe))
+  | where isnotempty(UserAgent)
+  | summarize by UserAgent, UserId
+  | join kind = rightanti (Dynamics365Activity
+  | where TimeGenerated > ago(timeframe)
+  | where isnotempty(UserAgent)
+  | where UserAgent !in (known_useragents)
```
Comment (shainw) [inline Detections/Dynamics365Activity/NewDynamicsUserAgent.yaml:30]: Maybe use !in~ just to make sure we don't miss on character case.

File: Detections/Dynamics365Activity/NewDynamicsUserAgent.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  | where UserAgent !startswith "PowerApps"
+  | where UserId matches regex user_accounts)
+  on UserAgent, UserId
+  // Uncomment this section to exclude user agents with a rendering engine, indicating browsers.
+  //| join kind = leftanti(
+  //Dynamics365Activity
+  //| where TimeGenerated between(ago(lookback)..ago(timeframe))
+  //| where UserAgent has_any ("Gecko", "WebKit", "Presto", "Trident", "EdgeHTML", "Blink")) on UserAgent
+  | summarize FirstSeen = min(TimeGenerated), IPs = make_set(ClientIP) by UserAgent, UserId
+  | extend AccountCustomEntity = UserId
```
Comment (shainw) [inline Detections/Dynamics365Activity/NewDynamicsUserAgent.yaml:41]: Still want to extend timestamp for now - timestamp = FirstSeen

File: Detections/Dynamics365Activity/NewOfficeUserAgentinDynamics.yaml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  - InitialAccess
+relevantTechniques:
+  - T1078
+query: |
+  let timeframe = 1h;
+  let lookback = 7d;
+  let known_useragents = dynamic([]);
+  Dynamics365Activity
+  | where TimeGenerated > ago(timeframe)
+  | extend Message = split(OriginalObjectId, ' ')[0]
```
Comment (shainw) [inline Detections/Dynamics365Activity/NewOfficeUserAgentinDynamics.yaml:24]: I think you need to cast tostring here.

File: Detections/Dynamics365Activity/NewOfficeUserAgentinDynamics.yaml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  | where Message =~ "UserSignIn"
+  | extend IPAddress = tostring(split(ClientIP, ":")[0])
+  | where isnotempty(UserAgent)
+  // Exclude user agents with a render agent to reduce noise
+  | where UserAgent has_any ("Gecko", "WebKit", "Presto", "Trident", "EdgeHTML", "Blink")
+  | join kind=leftanti(
+  OfficeActivity
+  | where TimeGenerated > ago(lookback)
+  | where UserAgent !in (known_useragents))
```
Comment (shainw) [inline Detections/Dynamics365Activity/NewOfficeUserAgentinDynamics.yaml:33]: !in~ here also?

File: Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+tactics:
+  - Collection
+relevantTechniques:
+  - T1530
+query: |
+  let baseline_time = 14d;
+  let detection_time = 1d;
+  Dynamics365Activity
+  | where TimeGenerated between(ago(baseline_time)..ago(detection_time-1d))
+  | extend Message = split(OriginalObjectId, ' ')[0]
```
Comment (shainw) [inline Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml:23]: tostring()?

File: Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  | where Message =~ "RetrieveMultiple"
+  | extend numQueryCount = todouble(QueryResults)
+  | extend QueryCount = iif(QueryResults contains ",", todouble(countof(tostring(QueryResults), ',') + 1), numQueryCount)
```
Comment (shainw) [inline Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml:26]: array_length()?

File: Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  | extend QueryCount = iif(isnotempty(QueryCount), QueryCount, double(1))
+  | summarize sum(QueryCount) by UserId
+  | extend HistoricalBaseline = sum_QueryCount
+  | join (Dynamics365Activity
+  | where TimeGenerated > ago(detection_time)
+  | extend Message = split(OriginalObjectId, ' ')[0]
+  | where Message =~ "RetrieveMultiple"
+  | extend numQueryCount = todouble(QueryResults)
+  | extend QueryCount = iif(QueryResults contains ",", todouble(countof(tostring(QueryResults), ',') + 1), numQueryCount)
```
Comment (shainw) [inline Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml:35]: array_length()?

File: Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  | extend QueryCount = iif(isnotempty(QueryCount), QueryCount, double(1))
+  | summarize sum(QueryCount) by UserId
+  | extend HistoricalBaseline = sum_QueryCount
+  | join (Dynamics365Activity
+  | where TimeGenerated > ago(detection_time)
+  | extend Message = split(OriginalObjectId, ' ')[0]
```
Comment (shainw) [inline Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml:32]: tostring()?

File: Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  | extend numQueryCount = todouble(QueryResults)
+  | extend QueryCount = iif(QueryResults contains ",", todouble(countof(tostring(QueryResults), ',') + 1), numQueryCount)
+  | extend QueryCount = iif(isnotempty(QueryCount), QueryCount, double(1))
+  | summarize sum(QueryCount) by UserId
+  | extend CurrentExportRate = sum_QueryCount) on UserId
+  | where CurrentExportRate > HistoricalBaseline
+  | project UserId, HistoricalBaseline, CurrentExportRate
+  | join kind=inner(Dynamics365Activity
+  | where TimeGenerated > ago(detection_time)
+  | extend Message = split(OriginalObjectId, ' ')[0]
```
Comment (shainw) [inline Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml:43]: tostring()?

File: Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+  | where Message =~ "RetrieveMultiple"
+  | extend numQueryCount = todouble(QueryResults)
+  | extend QueryCount = iif(QueryResults contains ",", todouble(countof(tostring(QueryResults), ',') + 1), numQueryCount)
```
Comment (shainw) [inline Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml:46]: array_length()?

File: Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+  | extend QueryCount = iif(isnotempty(QueryCount), QueryCount, double(1))) on UserId
+  | project TimeGenerated, UserId, QueryCount, UserAgent, Message, ClientIP, HistoricalBaseline, CurrentExportRate, CorrelationId, CrmOrganizationUniqueName, Query
+  | summarize QuerySizes = make_set(QueryCount), MostRecentQuery = max(TimeGenerated), IPs = make_set(ClientIP), UserAgents = make_set(UserAgent), make_set(Query) by UserId, CrmOrganizationUniqueName, HistoricalBaseline, CurrentExportRate
+  | extend AccountCustomEntity = UserId
```
Comment (shainw) [inline Detections/Dynamics365Activity/UserBulkRetreivalOutsideNormalActivity.yaml:50]: timestamp = MostRecentQuery

**Review by shainw — CHANGES_REQUESTED** [review]
Some fixes needed.

File: Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml (hunk 2 above)
Comment (petebryan) [reply Detections/Dynamics365Activity/MassExportOfDynamicstoExcel.yaml:25]: It is a dynamic but not in that way. The type depends on the size of query, sometimes its a int, sometimes a double, and sometimes a string of ids. This query is very clunkly but unfortunately its needed due to the data.

File: Detections/Dynamics365Activity/NewOfficeUserAgentinDynamics.yaml (hunk 8 above)
Comment (petebryan) [reply Detections/Dynamics365Activity/NewOfficeUserAgentinDynamics.yaml:24]: On what element here? OriginalObjectId is string type already (do you mean make Message string?

**Review by shainw — APPROVED** [review]

**Pull request merged**